### PR TITLE
8311645: Memory leak in jspawnhelper spawnChild after JDK-8307990

### PR DESCRIPTION
**Pull request by kspeeyu**
Free-ing 'buf' before two conditional return statements introduced by JDK-8307990

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311645](https://bugs.openjdk.org/browse/JDK-8311645): Memory leak in jspawnhelper spawnChild after JDK-8307990 (**Bug** - P3)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Volker Simonis](https://openjdk.org/census#simonis) (@simonis - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14809/head:pull/14809` \
`$ git checkout pull/14809`

Update a local copy of the PR: \
`$ git checkout pull/14809` \
`$ git pull https://git.openjdk.org/jdk.git pull/14809/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14809`

View PR using the GUI difftool: \
`$ git pr show -t 14809`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14809.diff">https://git.openjdk.org/jdk/pull/14809.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14809#issuecomment-1629483587)

### CODE REVIEW

File: src/java.base/unix/native/libjava/ProcessImpl_md.c [hunk 1]
```diff
@@ -562,15 +562,17 @@ spawnChild(JNIEnv *env, jobject process, ChildStuff *c, const char *helperpath)
 
     /* write the two structs and the data buffer */
     if (writeFully(c->childenv[1], (char *)&magic, sizeof(magic)) != sizeof(magic)) { // magic number first
-        return -1;
+        free(buf);
+	return -1;
```
Comment (shipilev) [inline src/java.base/unix/native/libjava/ProcessImpl_md.c:566]: See jcheck warnings: there is `TAB` on this line, which should be replaced by 4-space indents.

File: src/java.base/unix/native/libjava/ProcessImpl_md.c [hunk 2]
```diff
@@ -562,15 +562,17 @@ spawnChild(JNIEnv *env, jobject process, ChildStuff *c, const char *helperpath)
...
     }
 #ifdef DEBUG
     jtregSimulateCrash(resultPid, 2);
 #endif
     if (writeFully(c->childenv[1], (char *)c, sizeof(*c)) != sizeof(*c) ||
         writeFully(c->childenv[1], (char *)&sp, sizeof(sp)) != sizeof(sp) ||
         writeFully(c->childenv[1], buf, bufsize) != bufsize) {
-        return -1;
+        free(buf);
+	return -1;
```
Comment (shipilev) [inline src/java.base/unix/native/libjava/ProcessImpl_md.c:575]: Same here.

**Review by shipilev — CHANGES_REQUESTED** [review]
Looks okay, but need to fix two whitespace errors for jcheck to be happy.
Also, attention @simonis, whose patch introduced this little problem, I think.

File: src/java.base/unix/native/libjava/ProcessImpl_md.c (hunk 2 above)
Comment (kspeeyu) [reply src/java.base/unix/native/libjava/ProcessImpl_md.c:575]: Thanks for that. Will do

**Comment by simonis** [comment]
Hi @kspeeyu ,

Thanks for fixing this issue. The fix looks good except the whitespace issue already mentioned before.

Notice that we don't use TABs for code indentation in OpenJDK. For the native Hotspot sources (i.e. the ones under `src/hotspot/`) we use an indentation of two spaces whereas for Java code and the native part of the class library (i.e. the native code outside `src/hotspot`) we use four spaces. We don't change the indentation of imported libraries like HarfBuzz or zlib at all.

You can find more details on the coding style under https://github.com/openjdk/jdk/tree/master/doc. Just in case you haven't seen it yet, the [OpenJDK Developers’ Guide](https://openjdk.org/guide/) also contains a lot of useful information.

**Review by shipilev — APPROVED** [review]
Looks good!

**Comment by kspeeyu** [comment]
@simonis, thanks for your response. Fixed the whitespace, will follow the coding style guidelines.

**Review by simonis — APPROVED** [review]
Thanks, looks good now.

I can sponsor your change once you `/integrate`.

**Comment by simonis** [comment]
> Hold on a sec, let's make sure the GitHub Actions complete with all green. We can techinically wait after `/integrate` is said for sponsored changeset, but it is cleaner process-wise to wait for GHA all-clear/all-green.

Sure thing :)

**Review by AlanBateman — APPROVED** [review]

**Comment by shipilev** [comment]
GHA reports one test failure, but it looks infrastructural. Feel free to say `/integrate`, @kspeeyu.

**Review by RogerRiggs — APPROVED** [review]

**Pull request closed without merging**